From 5f99fb2c2b55acda2dee285f9bc3d135b477a1a0 Mon Sep 17 00:00:00 2001
From: Simon McVittie <smcv@collabora.com>
Date: Wed, 15 Aug 2018 16:21:52 +0100
Subject: [PATCH] Restore ability to boot pre-stretch systemd systems

The systemctl in Debian jessie didn't support the --no-block option.

Signed-off-by: Simon McVittie <smcv@collabora.com>
---
 lib/VirtSubproc.py | 32 ++++++++++++++++++++++----------
 1 file changed, 22 insertions(+), 10 deletions(-)

diff --git a/lib/VirtSubproc.py b/lib/VirtSubproc.py
index 671a21d..847eac8 100644
--- a/lib/VirtSubproc.py
+++ b/lib/VirtSubproc.py
@@ -309,6 +309,27 @@ def cmd_revert(c, ce):
     return [downtmp]
 
 
+def reboot_testbed():
+    (systemd_check, out, err) = execute_timeout(
+        None, 10, auxverb + ['test', '-d', '/run/systemd/system'],
+        stdout=subprocess.PIPE, stderr=subprocess.PIPE
+    )
+    if systemd_check == 0:
+        reboot_cmd = ['systemd-run', '--no-block', '--quiet', 'sh', '-c', 'sleep 3; reboot']
+        adtlog.debug('rebooting with command: %s' % reboot_cmd)
+        status, out, err = execute_timeout(None, 30, auxverb + reboot_cmd)
+
+        if status == 0:
+            return
+        else:
+            adtlog.debug(
+                'Cannot reboot with %s, trying something else' % reboot_cmd)
+
+    reboot_cmd = ['sh', '-c', '(sleep 3; reboot) >/dev/null 2>&1 &']
+    adtlog.debug('rebooting with command: %s' % reboot_cmd)
+    status, out, err = execute_timeout(None, 30, auxverb + reboot_cmd)
+
+
 def cmd_reboot(c, ce):
     global downtmp
     cmdnumargs(c, ce, 0, 1)
@@ -337,17 +358,8 @@ def cmd_reboot(c, ce):
     if len(c) > 1 and c[1] == 'prepare-only':
         adtlog.info('state saved, waiting for testbed to reboot...')
     else:
-        (systemd_check, out, err) = execute_timeout(
-            None, 10, auxverb + ['test', '-d', '/run/systemd/system'],
-            stdout=subprocess.PIPE, stderr=subprocess.PIPE
-        )
-        if systemd_check == 0:
-            reboot_cmd = ['systemd-run', '--no-block', '--quiet', 'sh', '-c', 'sleep 3; reboot']
-        else:
-            reboot_cmd = ['sh', '-c', '(sleep 3; reboot) >/dev/null 2>&1 &']
+        reboot_testbed()
 
-        adtlog.debug('rebooting with command: %s' % reboot_cmd)
-        execute_timeout(None, 30, auxverb + reboot_cmd)
     caller.hook_wait_reboot()
 
     # restore downtmp
-- 
2.18.0

